In [451]:
import pandas as pd
def split_count(x):
    '''
     Returns a dataframe object created from the input which is a series object 
    '''
    assert type(x) is pd.Series
    assert len(x)>0
    
    split_dict = {}
    for i in x:
        for key_of_dict in i.split(','):
            key_of_dict = key_of_dict.strip()
            if key_of_dict in split_dict.keys() :
                split_dict[key_of_dict] +=1
            else:
                split_dict[key_of_dict] =1
    df = pd.DataFrame.from_dict(split_dict, orient='index', columns = ['count'])
    
    return df
    

In [452]:
x = pd.read_csv('survey_data.csv')


In [453]:
split_count(input_)

count
Data extraction and processing             291
Data analytics                             351
Machine learning                           381
Statistical analysis                       222
Numerical processing                       173
All of the above                             1
Scientific visualization                   198
Web development                            104
Computer vision                              1
Image Processing                             1
Computer vision/image processing             1
As a general skill                           1
scripting seems desirable for many jobs      1
not sure                                     1
Computer Vision                              1
EDA tools                                    1

In [454]:
import pandas as pd

def add_month_yr(x):
    '''
     Returns dataframe x with an additional column month_yr added to it
    '''
    assert type(x) is pd.DataFrame 
    assert len(x)>0
    assert all(isinstance(i, str)for i in x['Timestamp'])
    
    col_to_add = pd.to_datetime(x['Timestamp'])
    year = col_to_add.dt.to_period('Y')
    month = col_to_add.dt.strftime('%b')

    x['month-yr'] = month + '-' + year.map(str)
    
    return x

In [455]:
p = add_month_yr(x)

In [456]:
def count_month_yr(x):
    '''
     Returns a dataframe which has counts of values of the column 'month-yr'
    '''
    assert type(x) is pd.DataFrame
    assert len(x)>0
    
    new_df = add_month_yr(x)
    
    return pd.DataFrame(new_df['month-yr'].value_counts())

In [457]:
count_month_yr(x)

month-yr
Jan-2018       148
Sep-2018       130
Sep-2017        74
Jan-2019        57
Mar-2018        41
Apr-2018        28
Oct-2018         6
Feb-2018         2

The problem with our previous result is that the order is wrong. Convert the month-yr column dtype to a Pandas CategoricalDtype with the correct order. You should be able to reproduce the following statement,

>>> x.groupby('month-yr')['Timestamp'].count().to_frame().sort_index() 
Timestamp
month-yr	
Sep-2017	74
Jan-2018	148
Feb-2018	2
Mar-2018	41
Apr-2018	28
Sep-2018	130
Oct-2018	6
Jan-2019	57
Note that the groupby is now sorted correctly. Your function signature is fix_categorical(x). It should take the month-yr dataframe column and then return the same dataframe with an updated column of CategoricalDtype that does the sorting as described. Remember to include your add_month_year code from the previous part, as your new function needs the output from it.

Here is your function signature fix_categorical(x) where x is a pd.DataFrame with the required "month-yr" column and output is a pd.DataFrame with the "month-yr" column having the categorical dtype.

Please put your Python code in a Python script file and upload it. Please retain your submitted source files! Remember to use all the best practices we discussed in class. You can use any module in the Python standard library, but third-party modules (e.g., Numpy, Pandas) are restricted to those explicitly mentioned in the problem description.

After you have submitted your file, do not use the browser back or reload buttons to navigate or open the page in multiple browser tabs, as this may cause your attempts to decrease unexpectedly. It may take up to thirty seconds for your code to be processed, so please be patient.

Good luck!

In [530]:
import pandas as pd
import calendar

def add_month_yr(x):
    '''
     Returns dataframe x with an additional column month_yr added to it
    '''
    assert type(x) is pd.DataFrame 
    assert len(x)>0
    assert all(isinstance(i, str)for i in x['Timestamp'])
    
    col_to_add = pd.to_datetime(x['Timestamp'])
    year = col_to_add.dt.to_period('Y')
    month = col_to_add.dt.strftime('%b')

    x['month-yr'] = month + '-' + year.map(str)
    
    return x

def fix_categorical(x):
    '''
     Returns a dataframe 
    '''
    assert type(x) is pd.DataFrame
    assert len(x)>0
    assert 'month-yr' in x.columns
    assert 'Timestamp' in x.columns
    
    month_list = list(calendar.month_abbr)
    #print(month_list)
    month_yr_unique = list(set(x['month-yr']))
    #print(month_yr_unique)
    sort_month_yr =  lambda x:x.split('-')[1]+'{0:02d}'.format(month_list.index(x.split('-')[0]))
    month_yr_unique.sort(key = sort_month_yr)
    cats = pd.CategoricalDtype(month_yr_unique, ordered = True) 
    x['month-yr'] = x['month-yr'].astype({'month-yr':cats})
    return x
    

In [531]:
#df.groupby('user_created_at_year_month').size().reset_index().rename(columns={0: 'count_signups'})
z = fix_categorical(x)
#z['month-yr'].value_counts()

In [532]:
z.groupby('month-yr')['Timestamp'].count().to_frame().sort_index() 

Timestamp
month-yr           
Sep-2017         74
Jan-2018        148
Feb-2018          2
Mar-2018         41
Apr-2018         28
Sep-2018        130
Oct-2018          6
Jan-2019         57

In [491]:
l =p['month-yr']

In [ ]:
import numpy as np
cats = pd.CategoricalDtype(categories = l.unique(), ordered = True)  
p['month-yr'] =  pd.Series(p['month-yr'], dtype = cats)
#print(p['month-yr'])
# l.value_counts
#cats.categories
#list_cat =
#len(a)
#cats
# print(type(l))

p['month-yr'].dort

In [ ]:
a = ['one', 'two', 'two']
bins = 2
cats=pd.cut(a,bins,labels=['one','two'])
cats